### PR TITLE
stm32: add lvgl binding

### PR DESCRIPTION
**Pull request by cidermole**
add LV_ROOTS to MICROPY_PORT_ROOT_POINTERS for ports/stm32, just like it was done in the esp32 port

### CODE REVIEW

**Comment by amirgon** [comment]
Hi @cidermole,
Thank you for your contribution!

- On ESP32 I'm registering another module `lvesp32` which calls periodically to `lv_tick_inc` and `lv_task_handler`.  These functions must be called in order for lvgl to work correctly.
How do you call `lv_tick_inc` and `lv_task_handler` on stm32?
- What display/input devices do you use? How do you register them?

**Comment by cidermole** [comment]
Hi @amirgon,

* I just call them from a mainloop in `main.py`, like this (**Variant 1**):
```
import lvgl as lv

while True:
    lcd_v3.update()  # has a ~ 17 ms delay
    lv.tick_inc(17)
    lv.task_handler()
```
**Variant 2**: in theory, it should work with a timer in `boot.py`, e.g.
```
import pyb
import lvgl as lv

def tick(timer):
    lv.tick_inc(10)  # called every 10 ms
    lv.task_handler()
timer = pyb.Timer(4, freq=100)    # timer 4, trigger at 100 Hz
timer.callback(tick)
```
In practice, memory is very tight for me - I use pyboard v1.1 (192 KB total RAM) and have an ~58 kB bytearray as framebuffer. Using variant 2, I got an "uncaught exception in Timer(4) interrupt handler: MemoryError", and so I use variant 1.

* I'm using a reverse-engineered display that I drive using some bit-banging assembly in `lcd_v3.write()`. It gets the address of the framebuffer and writes out the entire screen contents (which is how the display is driven in its original application; I'm not sure if it can do partial updates at all). Initialization in `boot.py` looks like this:

```
# should start with this, otherwise there may not be enough memory left
frame_buf = bytearray(240*240 + 240 + 9*4)

import lvgl as lv
lv.init()

import lcd_v3

def my_flush_cb(disp_drv, area, color_p):
    lcd_v3.write(frame_buf)

# Display buffer
disp_buf = lv.disp_buf_t()
lv.disp_buf_init(disp_buf, frame_buf, None, 240*240)

# Display driver
disp_drv = lv.disp_drv_t()
lv.disp_drv_init(disp_drv)
disp_drv.buffer = disp_buf
disp_drv.flush_cb = my_flush_cb
disp_drv.hor_res = 240
disp_drv.ver_res = 240
disp = lv.disp_drv_register(disp_drv)
```

**Comment by embeddedt** [comment]
Don't call `lv_tick_inc` and `lv_task_handler` in the same loop. That has led to issues in the past. `lv_tick_inc` can safely be called while `lv_task_handler` is running, so it's better to call it in a timer interrupt (or whatever the Python equivalent is).

**Comment by cidermole** [comment]
Thanks! **Variant 3** which I now use:
```
# boot.py

def tick(timer):
    lv.tick_inc(10)  # called every 10 ms
timer = pyb.Timer(4, freq=100)    # timer 4, trigger at 100 Hz
timer.callback(tick)
```

```
# main.py
while True:
    lv.task_handler()
    # do other stuff...
```

Okay, so it would be great to create a module like `lvesp32` for stm32. 
Since Timer callbacks are executed in an interrupt context, it's probably not a good idea to call `lv_task_handler()` in there. 
Also, to avoid hard-wiring a timer and potentially collide with the user's choice, I think the best way for stm32 would be to pass a timer to an init function from Python.
 
Something like `import lvstm32; lvstm32.init(pyb.Timer(4))`, which would then internally call `timer.init(freq=100)` and register a timer callback implemented in C that calls `mp_sched_schedule()` to defer the `lv_task_handler()`. 
I'll need to wrap my head around how to write that C module though, lots of Python internals there I'm not very used to.

**Comment by amirgon** [comment]
> In practice, memory is very tight for me - I use pyboard v1.1 (192 KB total RAM)

In cases where memory is tight, you can disable features on both lvgl and Micropython (on `lv_conf.h` and `mpconfigport.h`). Especially try to remove fonts and themes. This could be a big memory saver. When disabling widgets and features in lvgl, the Micropython binding code itself would get smaller too and only provide API for the enabled features.

@kisvegabor / @embeddedt - this is another case where conditional defined on `lv_conf.h` could be useful, as [discussed](https://github.com/littlevgl/lvgl/issues/1243#issuecomment-560019864).

> * I'm using a reverse-engineered display that I drive using some bit-banging assembly in `lcd_v3.write()`

That sounds like an interesting project!  
Most users use a standard display controller such as ILI9341. Did you already write about your project somewhere?  
If you want, you can write a post in [My Projects section on lvgl forum](https://forum.littlevgl.com/c/my-projects), I'm sure this would be interesting for others.

> Using variant 2, I got an "uncaught exception in Timer(4) interrupt handler: MemoryError", and so I use variant 1.

Maybe because you allocated memory in an interrupt handler, which is forbidden, but can be worked around.
Have a look at https://docs.micropython.org/en/latest/reference/isr_rules.html

> Okay, so it would be great to create a module like `lvesp32` for stm32.

Do you use FreeRTOS or some other RTOS with stm32? If you do, it should be pretty simple to port lvesp32 to stm32, it's a very simple driver, and it doesn't really depend on any ESP32 feature beside registering a timer.  


> Something like `import lvstm32; lvstm32.init(pyb.Timer(4))`, which would then internally call `timer.init(freq=100)` and register a timer callback implemented in C that calls `mp_sched_schedule()` to defer the `lv_task_handler()`.
> I'll need to wrap my head around how to write that C module though, lots of Python internals there I'm not very used to.

In fact, something like lvesp32 could be implemented in Micropython, saving you the need to write a C module.  
The idea is that the timer interrupt calls `lv_tick_inc` (it's allowed to call it in interrupt context), and schedules a call to `lv_task_handler` which can't be called in interrupt context.
You can do both directly from Micropython, but you need to be careful not to allocate memory in ISR.

Something like:

```python
class lvstm32():
    def __init__(self, freq=25):
        self.task_handler_ref = self.task_handler  # Allocation occurs here
        self.delay = 1000 // freq
        tim = pyb.Timer(4)
        tim.init(freq=freq)
        tim.callback(self.timer_cb)

    def task_handler(self, _):
        lv.task_handler()

    def timer_cb(self, t):
        lv.tick_inc(self.delay)
        # Passing self.task_handler would cause allocation.
        micropython.schedule(self.task_handler_ref, 0)
```

**Comment by kisvegabor** [comment]
@kisvegabor / @embeddedt - this is another case where conditional defines on lv_conf.h could be useful, as discussed.

I'd still prefer to add `#ifdef`s to the `lv_conf.h` of `lv_micropython` locally to keep the `lv_conf.h` more readable.

**Comment by amirgon** [comment]
@cidermole Did you get a chance to try my suggestions above?
I'd love to add stm32 support to lv_micropython, but I can't check it so I need some help from someone who is actually using stm32.

File: ports/stm32/modules/lvstm32.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Module for advancing tick count and scheduling async event for lvgl on STM32.
+# Import after lvgl module.
+#
+# MIT license; Copyright (c) 2019 Amir Gonnen
```
Comment (embeddedt) [inline ports/stm32/modules/lvstm32.py]: Maybe we should put 2020 here since it's a new file?

**Comment by cidermole** [comment]
@amirgon Thanks, your class lvstm32 from above works perfectly for me. I've added it to the PR.

**Comment by amirgon** [comment]
> @amirgon Thanks, your class lvstm32 from above works perfectly for me. I've added it to the PR.

@cidermole Thanks for checking this.
Your branch has conflicts. Could you sync it to latest version and resolve them?

**Comment by cidermole** [comment]
@amirgon can you check if esp32 still builds with the added `strncpy()` in https://github.com/littlevgl/lv_micropython/pull/16/files#diff-cfdb30a8f0d8987b1d0e5d6a26fd838b? Somehow ports/stm32 is missing that.

**Comment by embeddedt** [comment]
@cidermole I'm not sure that you should modify MicroPython itself; this fork is just supposed to add LittlevGL-specific things.

**Comment by cidermole** [comment]
@embeddedt okay, removed `strncpy()` from here.

**Comment by amirgon** [comment]
Merged.
Thank you!

**Pull request merged**